examples/vision/deit.py

```diff
@@ -0,0 +1,575 @@
+"""
+Title: Distilling Vision Transformers
+Author: [Sayak Paul](https://twitter.com/RisingSayak)
+Date created: 2022/04/05
+Last modified: 2022/04/05
+Description: Distillation of Vision Transformers through attention.
+"""
+"""
+## Introduction
+
+In the original Vision Transformers (ViT) paper
+([Dosovitskiy et al.](https://arxiv.org/abs/2010.11929)),
+the authors concluded that to perform on par with Convolutional Neural Networks (CNN),
+ViTs need to be pre-trained on larger datasets. The larger the better. This is mainly
+due to the lack of inductive biases inside ViTs -- unlike CNNs they don't have layers
+that exploit locality. In a follow-up paper
+([Steiner et al.](https://arxiv.org/abs/2106.10270)),
+the authors show that it is possible to substantially improve the performance of ViTs
+with stronger regularization and longer training lengths.
+
+Many groups have proposed different ways to deal with the problem of data hunger of ViTs.
+One such way was shown in DeiT ([Touvron et al.](https://arxiv.org/abs/2012.12877)). The
+authors introduced a distillation technique that is specific to transformer-based vision
+models. DeiT is among the first works that showed it's possible to train ViTs well
+without using larger datasets. 
+
+In this example, we'll implement the distillation recipe proposed in DeiT. This will
+require us to slightly tweak the original ViT architecture and write a custom loop to
+implement the distillation recipe.
+
+To run the example, you'll need TensorFlow Addons which you can install with the
+following command:
+
+```
+pip install tensorflow-addons
+```
+
+To comfortably navigate through this example, you'll be expected to know how a ViT and
+knowledge distillation work. The following are good resources in case you needed a
+refresher:
+
+*
+[ViT on keras.io](https://keras.io/examples/vision/image_classification_with_vision_transformer) 
+* 
+[Knowledge distillation on keras.io](https://keras.io/examples/vision/knowledge_distillation/)
+"""
+
+"""
+## Imports
+"""
+
+from typing import List
+
+import tensorflow as tf
+import tensorflow_addons as tfa
+import tensorflow_datasets as tfds
+import tensorflow_hub as hub
+from tensorflow import keras
+from tensorflow.keras import layers
+
+tfds.disable_progress_bar()
+tf.keras.utils.set_random_seed(42)
+
+"""
+## Constants
+"""
+
+# Model
+MODEL_TYPE = "deit_distilled_tiny_patch16_224"
+RESOLUTION = 224
+PATCH_SIZE = 16
+NUM_PATCHES = (RESOLUTION // PATCH_SIZE) ** 2
+LAYER_NORM_EPS = 1e-6
+PROJECTION_DIM = 192
+NUM_HEADS = 3
+NUM_LAYERS = 12
+MLP_UNITS = [
+    PROJECTION_DIM * 4,
+    PROJECTION_DIM,
+]
+DROPOUT_RATE = 0.0
+DROP_PATH_RATE = 0.1
+
+# Training
+NUM_EPOCHS = 20
+BASE_LR = 0.0005
+WEIGHT_DECAY = 0.0001
+
+# Data
+BATCH_SIZE = 256
+AUTO = tf.data.AUTOTUNE
+NB_CLASSES = 5
```

Comment:
Nit: prefer `num_` for the number prefix for general consistency (as you do elsewhere in this example).